Cache dbuf_hash() calculation

ryao opened the pull request:
### Motivation and Context
We currently compute a 64-bit hash three times, which consumes 0.8% CPU time on ARC eviction heavy workloads. 

This was done by Klara Systems and sponsored by Wasabi Technology, Inc.

### Description
We cache the 64-bit hash.

### How Has This Been Tested?
It has been build tested and runtime tested, a couple months ago on an older version of the code.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

adamdmoss commented:
I have a small worry, which is that this may be intentional (or 67% intentional...) to catch memory corruption in-place.  But probably not.

ryao commented:
> I have a small worry, which is that this may be intentional (or 67% intentional...) to catch memory corruption in-place. But probably not.

Recalculating the hash value for dbuf_create after dbuf_find already calculated it and failed to find it would not help to catch corruption in place.

Recalculating it following a bit flip when removing the dbuf would cause us to fail to remove the dbuf.

I do not believe that the recalculation hardens us against bad memory.

ahrens commented on module/zfs/dbuf.c:
```diff
@@ -400,7 +403,7 @@ dbuf_hash_insert(dmu_buf_impl_t *db)
 	blkid = db->db_blkid;
-	hv = dbuf_hash(os, obj, level, blkid);
+	hv = db->db_hv;
 	idx = hv & h->hash_table_mask;
```
looks like you could inline `hv` into this one use:
```c
idx = dv->db_hv & h->hash_table_mask;
```

ryao replied:
Done.

ahrens commented on module/zfs/dbuf.c:
```diff
@@ -400,7 +403,7 @@ dbuf_hash_insert(dmu_buf_impl_t *db)
 
 	blkid = db->db_blkid;
-	hv = dbuf_hash(os, obj, level, blkid);
+	hv = db->db_hv;
```
maybe assert that the dbuf_hv still matches dbuf_hash(...)?

ryao replied:
Done.

ahrens commented on module/zfs/dbuf.c:
```diff
@@ -478,8 +481,7 @@ dbuf_hash_remove(dmu_buf_impl_t *db)
 
-	hv = dbuf_hash(db->db_objset, db->db.db_object,
-	    db->db_level, db->db_blkid);
+	hv = db->db_hv;
```
same here - inline hv and assert that it matches dbuf_hash().

ryao commented:
I plan to revise this some time on Wednesday.

ryao commented on module/zfs/dbuf.c:
```diff
@@ -3073,7 +3076,7 @@ dbuf_findbp(dnode_t *dn, int level, uint64_t blkid, int fail_sparse,
 static dmu_buf_impl_t *
 dbuf_create(dnode_t *dn, uint8_t level, uint64_t blkid,
-    dmu_buf_impl_t *parent, blkptr_t *blkptr)
+    dmu_buf_impl_t *parent, blkptr_t *blkptr, uint64_t *hash)
```
Would changing the naming to `hash_in` and `hash_out` in the different functions be sufficient?

ryao replied:
On second thought, requiring it as a parameter to `dbuf_create()` seems fine to me. I just need to add code to `dbuf_create_bonus()` to compute it. I like that better. I have pushed a revision doing this. It also changes `dbuf_find()` to rename the pointer to `hash_out` to make the direction clear.

That should address this.

ryao commented on include/sys/dbuf.h:
```diff
@@ -294,6 +294,8 @@ typedef struct dmu_buf_impl {
 	/* Tells us which dbuf cache this dbuf is in, if any */
 	dbuf_cached_state_t db_caching_status;
 
+	uint64_t db_hv;
```
I picked `->db_hash`.

ryao commented:
Sorry for the delay. I was not feeling that great last week. Anyway, I addressed all of the comments. I just need to hear back from @ahrens on whether naming the pointer values `hash_in` and `hash_out` to designate the meaning is alright.

ryao commented:
> Sorry for the delay. I was not feeling that great last week. Anyway, I addressed all of the comments. I just need to hear back from @ahrens on whether naming the pointer values `hash_in` and `hash_out` to designate the meaning is alright.

On second thought, pulling out the hash calculation from `dbuf_create()` so we can unconditionally pass by value instead of optionally by pointer is better. I just did that and repushed.

ryao commented:
I did another push to fix a cstyle.pl complaint that I had missed. The change is purely a formatting fix.
